File: src/SharpOMatic.FrontEnd/src/app/pages/metrics/interfaces/model-call-metrics-dashboard.ts
export enum ModelCallMetricScope {
  All = 0,
  Workflow = 1,
  Connector = 2,
  Model = 3,
}

export enum ModelCallMetricBucket {
  Hour = 0,
  Day = 1,
  Week = 2,
}

export interface ModelCallMetricsDashboard {
  start: string;
  end: string;
  bucket: ModelCallMetricBucket;
  scope: ModelCallMetricScope;
  scopeKey: string | null;
  scopeName: string | null;
  totals: ModelCallMetricTotals;
  masterItems: ModelCallMetricMasterItem[];
  timeBuckets: ModelCallMetricTimeBucket[];
  workflowBreakdown: ModelCallMetricBreakdownItem[];
  connectorBreakdown: ModelCallMetricBreakdownItem[];
  modelBreakdown: ModelCallMetricBreakdownItem[];
  nodeBreakdown: ModelCallMetricBreakdownItem[];
  failures: ModelCallMetricFailureGroup[];
  failureCategories: ModelCallMetricFailureCategoryGroup[];
  recentCalls: ModelCallMetricCallSummary[];
  recentCallsTotal: number;
  slowestCalls: ModelCallMetricCallSummary[];
  recentLogicalCalls: ModelCallMetricLogicalCallSummary[];
  recentLogicalCallsTotal: number;
  slowestLogicalCalls: ModelCallMetricLogicalCallSummary[];
}

export interface ModelCallMetricTotals {
  totalCalls: number;
  successfulCalls: number;
  failedCalls: number;
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  totalCost: number;
  pricedCalls: number;
  unpricedCalls: number;
  averageDuration: number | null;
  p95Duration: number | null;
  failureRate: number;
  logicalCalls: number;
  callsRequiringFallback: number;
  recoveredCalls: number;
  unrecoveredCalls: number;
  fallbackRecoveryRate: number;
  logicalFailureRate: number;
}

export interface ModelCallMetricMasterItem {
  key: string;
  name: string;
  totalCalls: number;
  failedCalls: number;
  totalCost: number;
  totalTokens: number;
  averageDuration: number | null;
  p95Duration: number | null;
  failureRate: number;
}

export interface ModelCallMetricTimeBucket {
  start: string;
  totalCalls: number;
  successfulCalls: number;
  failedCalls: number;
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  totalCost: number;
  pricedCalls: number;
  unpricedCalls: number;
  averageDuration: number | null;
  p95Duration: number | null;
  logicalCalls: number;
  primaryAttempts: number;
  fallbackAttempts: number;
  successfulFallbackAttempts: number;
  failedFallbackAttempts: number;
  recoveredCalls: number;
  unrecoveredCalls: number;
}

export interface ModelCallMetricBreakdownItem {
  key: string;
  name: string;
  totalCalls: number;
  failedCalls: number;
  totalCost: number;
  totalTokens: number;
  averageDuration: number | null;
  p95Duration: number | null;
  failureRate: number;
  primaryAttempts: number;
  fallbackAttempts: number;
  successfulFallbackAttempts: number;
}

export interface ModelCallMetricFailureCategoryGroup {
  category: ModelFallbackFailureCategory;
  providerStatusCode: number | null;
  count: number;
  lastSeen: string;
  primaryAttempts: number;
  fallbackAttempts: number;
}

export enum ModelFallbackFailureCategory {
  Unknown = 0,
  RateLimited = 1,
  ProviderUnavailable = 2,
  Timeout = 3,
  Network = 4,
  Authentication = 5,
  InvalidRequest = 6,
  Configuration = 7,
  Cancellation = 8,
}

export interface ModelCallMetricFailureGroup {
  errorType: string;
  errorMessage: string;
  count: number;
  firstSeen: string;
  lastSeen: string;
  workflowCount: number;
  connectorCount: number;
  modelCount: number;
}

export interface ModelCallMetricCallSummary {
  id: string;
  logicalCallId: string;
  attemptNumber: number;
  created: string;
  workflowName: string;
  nodeTitle: string;
  connectorName: string | null;
  modelName: string | null;
  duration: number | null;
  inputTokens: number | null;
  outputTokens: number | null;
  totalTokens: number | null;
  totalCost: number | null;
  succeeded: boolean;
  failureCategory: number | null;
  providerStatusCode: number | null;
  errorType: string | null;
  errorMessage: string | null;
}

export interface ModelCallMetricLogicalCallSummary {
  logicalCallId: string;
  created: string;
  workflowName: string;
  nodeTitle: string;
  finalConnectorName: string | null;
  finalModelName: string | null;
  attemptCount: number;
  duration: number | null;
  totalTokens: number;
  totalCost: number | null;
  succeeded: boolean;
  recoveredByFallback: boolean;
  attempts: ModelCallMetricCallSummary[];
}

File: src/SharpOMatic.Engine/DTO/ModelCallMetricsDashboard.cs
namespace SharpOMatic.Engine.DTO;

public enum ModelCallMetricScope
{
    All = 0,
    Workflow = 1,
    Connector = 2,
    Model = 3,
}

public enum ModelCallMetricBucket
{
    Hour = 0,
    Day = 1,
    Week = 2,
}

public sealed record ModelCallMetricsDashboardRequest(
    DateTime Start,
    DateTime End,
    ModelCallMetricBucket Bucket,
    ModelCallMetricScope Scope,
    string? ScopeKey,
    string? MasterSearch,
    int RecentSkip,
    int RecentTake,
    bool AllTime = false
);

public sealed record ModelCallMetricsDashboard(
    DateTime Start,
    DateTime End,
    ModelCallMetricBucket Bucket,
    ModelCallMetricScope Scope,
    string? ScopeKey,
    string? ScopeName,
    ModelCallMetricTotals Totals,
    List<ModelCallMetricMasterItem> MasterItems,
    List<ModelCallMetricTimeBucket> TimeBuckets,
    List<ModelCallMetricBreakdownItem> WorkflowBreakdown,
    List<ModelCallMetricBreakdownItem> ConnectorBreakdown,
    List<ModelCallMetricBreakdownItem> ModelBreakdown,
    List<ModelCallMetricBreakdownItem> NodeBreakdown,
    List<ModelCallMetricFailureGroup> Failures,
    List<ModelCallMetricFailureCategoryGroup> FailureCategories,
    List<ModelCallMetricCallSummary> RecentCalls,
    int RecentCallsTotal,
    List<ModelCallMetricCallSummary> SlowestCalls,
    List<ModelCallMetricLogicalCallSummary> RecentLogicalCalls,
    int RecentLogicalCallsTotal,
    List<ModelCallMetricLogicalCallSummary> SlowestLogicalCalls
);

public sealed record ModelCallMetricTotals(
    int TotalCalls,
    int SuccessfulCalls,
    int FailedCalls,
    long InputTokens,
    long OutputTokens,
    long TotalTokens,
    decimal TotalCost,
    int PricedCalls,
    int UnpricedCalls,
    double? AverageDuration,
    long? P95Duration,
    double FailureRate,
    int LogicalCalls,
    int CallsRequiringFallback,
    int RecoveredCalls,
    int UnrecoveredCalls,
    double FallbackRecoveryRate,
    double LogicalFailureRate
);

public sealed record ModelCallMetricMasterItem(
    string Key,
    string Name,
    int TotalCalls,
    int FailedCalls,
    decimal TotalCost,
    long TotalTokens,
    double? AverageDuration,
    long? P95Duration,
    double FailureRate
);

public sealed record ModelCallMetricTimeBucket(
    DateTime Start,
    int TotalCalls,
    int SuccessfulCalls,
    int FailedCalls,
    long InputTokens,
    long OutputTokens,
    long TotalTokens,
    decimal TotalCost,
    int PricedCalls,
    int UnpricedCalls,
    double? AverageDuration,
    long? P95Duration,
    int LogicalCalls,
    int PrimaryAttempts,
    int FallbackAttempts,
    int SuccessfulFallbackAttempts,
    int FailedFallbackAttempts,
    int RecoveredCalls,
    int UnrecoveredCalls
);

public sealed record ModelCallMetricBreakdownItem(
    string Key,
    string Name,
    int TotalCalls,
    int FailedCalls,
    decimal TotalCost,
    long TotalTokens,
    double? AverageDuration,
    long? P95Duration,
    double FailureRate,
    int PrimaryAttempts,
    int FallbackAttempts,
    int SuccessfulFallbackAttempts
);

public sealed record ModelCallMetricFailureCategoryGroup(
    ModelFallbackFailureCategory Category,
    int? ProviderStatusCode,
    int Count,
    DateTime LastSeen,
    int PrimaryAttempts,
    int FallbackAttempts
);

public sealed record ModelCallMetricFailureGroup(
    string ErrorType,
    string ErrorMessage,
    int Count,
    DateTime FirstSeen,
    DateTime LastSeen,
    int WorkflowCount,
    int ConnectorCount,
    int ModelCount
);

public sealed record ModelCallMetricCallSummary(
    Guid Id,
    Guid LogicalCallId,
    int AttemptNumber,
    DateTime Created,
    string WorkflowName,
    string NodeTitle,
    string? ConnectorName,
    string? ModelName,
    long? Duration,
    long? InputTokens,
    long? OutputTokens,
    long? TotalTokens,
    decimal? TotalCost,
    bool Succeeded,
    ModelFallbackFailureCategory? FailureCategory,
    int? ProviderStatusCode,
    string? ErrorType,
    string? ErrorMessage
);

public sealed record ModelCallMetricLogicalCallSummary(
    Guid LogicalCallId,
    DateTime Created,
    string WorkflowName,
    string NodeTitle,
    string? FinalConnectorName,
    string? FinalModelName,
    int AttemptCount,
    long? Duration,
    long TotalTokens,
    decimal? TotalCost,
    bool Succeeded,
    bool RecoveredByFallback,
    List<ModelCallMetricCallSummary> Attempts
);
